### ISSUE TITLE
Add assertions for ES6 Maps and Sets

### ISSUE THREAD

**Issue by tiemevanveen** [chaijs/chai#632]
@keithamus:
> @tiemevanveen you should raise a new issue - we can discuss the design of this and work towards adding it as a feature :smile:

------------------ 

It would be cool if there would be assertions for ES6 Maps (and Sets).

For example something like:

    expect(myMap).to.have.key('some_key');

Currently one could use:

    expect(myMap.has('some_key'), 'map should have some_key').to.be.true;

But that doesn't work great..

**Comment by keithamus** [comment]
We already have a [`.keys` assertion](https://github.com/chaijs/chai/blob/master/lib/chai/core/assertions.js#L1115-L1235) so it makes sense to extend this to detect maps and sets. You would need to wrap the existing [key extraction code](https://github.com/chaijs/chai/blob/master/lib/chai/core/assertions.js#L1174-L1180) into the `else` of an `if` that type detects maps and sets. So the result would look a little like:

```js
var len = keys.length,
    any = flag(this, 'any'),
    all = flag(this, 'all'),
    actual,
    expected;
if (_.type(obj) === 'map') {
  actual = Array.from(obj.keys());
  expected = keys;
} else {
  actual = Object.keys(obj);
  expected = keys.map(String);
}
```

A good PR would also include some test coverage, and extend the existing documentation (found in the comments above the method).

**Comment by lucasfcosta** [comment]
Hi everyone,
As I was working on this I had a doubt about the type of keys: **on Sets and Maps we can have virtually anything as a key**, so are we going to accept these types to be passed as arguments?

TL;DR
Should this be valid?
```js
var myMap = new Map().set({}, 'value');
expect(myMap).to.have.key({});
```

**Comment by keithamus** [comment]
@lucasfcosta awesome! Look forward to the PR :smile:. If it is a breaking change, make it against `4.x.x` - but it seems as though it'd be a feature without any breakages, as trying to assert if a map has a key will currently only cause an error; so PRing against `master` sounds like a good idea.

**Comment by bnord01** [comment]
@lucasfcosta 
> Should this, for example, be allowed?
```javascript
var myMap = new Map().set({}, 'value').set({lol: 'lel'}, 'laugh');
expect(myMap).to.have.all.keys({}, {lol: 'lel'});
```

I would expect this to fail. Keys in maps and values in sets are by reference and checked that way using `has`
```javascript
var k1 = {}, k2={}, myMap = new Map().set(k1,1)
myMap.should.have.key(k1) // myMap.has(k1).should.be.true
myMap.should.not.have.key(k2) // myMap.has(k2).should.be.false
```

**Comment by lucasfcosta** [comment]
@bnord01 thanks for the explanation, I didn't know it. Makes total sense to me. :smiley:

**Comment by jdmarshall** [comment]
I'm not a fan of how keys() insist on an exact match instead of a subset match, but includes.keys() can match on subsets.

As implemented, .all. seems to be superfluous. It implies that some other check happens but as far as I can tell it has the same corner cases.

I'm commenting here only partially to register a complaint, and more so the next person doing a google search figures out why their tests are red.
